crates/walrus-service/src/node/blob_event_processor.rs

```diff
@@ -0,0 +1,311 @@
+                BackgroundEventProcessor::new(node.clone(), blob_sync_handler.clone(), rx);
+            workers.push(Arc::new(tokio::spawn(async move {
+                background_processor.run(worker_index).await;
+            })));
+        }
+
+        let sequential_processor = if num_workers == 0 {
+            // Create a sequential processor to process events sequentially if no background workers
+            // are configured.
+            let (_tx, rx) = mpsc::unbounded_channel();
+            Some(Arc::new(BackgroundEventProcessor::new(
+                node.clone(),
+                blob_sync_handler.clone(),
+                rx,
+            )))
+        } else {
+            None
+        };
+
+        Self {
+            node,
+            background_processor_senders: senders,
+            _background_processors: workers,
+            sequential_processor,
+        }
+    }
+
+    /// Processes a blob event.
+    pub async fn process_event(
+        &self,
+        event_handle: EventHandle,
+        blob_event: BlobEvent,
+    ) -> anyhow::Result<()> {
+        // Update the blob info based on the event.
+        // This processing must be sequential and cannot be parallelized, since there is logical
+        // dependency between events.
+        self.node
+            .storage
+            .update_blob_info(event_handle.index(), &blob_event)?;
+
+        if let BlobEvent::Registered(_) = &blob_event {
+            // Registered event is marked as complete immediately. We need to process registered
+            // events as fast as possible to catch up to the latest event in order to not miss
+            // blob sliver uploads.
+            //
+            // If we want to do this in parallel, we shouldn't mix registered event processing with
+            // certified event processing, as certified events take longer and can block following
+            // registered events.
+            monitored_scope::monitored_scope("ProcessEvent::BlobEvent::Registered");
+            event_handle.mark_as_complete();
+        } else {
+            // If there are no background workers, we use a sequential processor to process events
+            // sequentially.
+            if self.background_processor_senders.is_empty() {
+                assert!(self.sequential_processor.is_some());
+                self.sequential_processor
+                    .as_ref()
+                    .expect(
+                        "sequential processor must be configured when no background \
+                            workers are configured",
+                    )
+                    .process_event(event_handle, blob_event)
+                    .await?;
+            } else {
+                // We send the event to one of the workers to process in parallel.
+                // Note that in order to remain sequential processing for the same BlobID, we always
+                // send events for the same BlobID to the same worker.
```

Comment:
Praise: I like this.

Remark (maybe to add to the comment): We need to be careful if we ever want to dynamically adjust the number of workers. (Not saying that we should do that, just to be sure.)